rust/src/asn1/mod.rs

```diff
@@ -0,0 +1,447 @@
+            if res.is_some() {
+                return Ok(res);
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Checks a BerObject and subnodes against the Asn1 checks
+    fn check_object(obj: &BerObject, ad: &DetectAsn1Data) -> Option<Asn1Check> {
+        // oversize_length will check if a node has a length greater than
+        // the user supplied length
+        if let Some(oversize_length) = ad.oversize_length {
+            if obj.header.len > oversize_length as u64
+                || obj.content.as_slice().unwrap_or(&[]).len() > oversize_length as usize
+            {
+                return Some(Asn1Check::OversizeLength);
+            }
+        }
+
+        // bitstring_overflow check a malformed option where the number of bits
+        // to ignore is greater than the length decoded (in bits)
+        if ad.bitstring_overflow
+            && (obj.header.is_universal()
+                && obj.header.tag == BerTag::BitString
+                && obj.header.is_primitive())
+        {
+            if let BerObjectContent::BitString(bits, _v) = &obj.content {
+                if obj.header.len > 0
+                    && *bits as u64 > (obj.header.len.checked_mul(8).unwrap_or(std::u64::MAX))
+                {
+                    return Some(Asn1Check::BitstringOverflow);
+                }
+            }
+        }
+
+        // double_overflow checks a known issue that affects the MSASN1 library
+        // when decoding double/real types. If the encoding is ASCII,
+        // and the buffer is greater than 256, the array is overflown
+        if ad.double_overflow
+            && (obj.header.is_universal()
+                && obj.header.tag == BerTag::RealType
+                && obj.header.is_primitive())
+        {
+            if let Ok(data) = obj.content.as_slice() {
+                if obj.header.len > 0
+                    && !data.is_empty()
+                    && data[0] & 0xC0 == 0
+                    && (obj.header.len > 256 || data.len() > 256)
+                {
+                    return Some(Asn1Check::DoubleOverflow);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn from_slice(input: &'a [u8], ad: &DetectAsn1Data) -> Result<Asn1<'a>, Asn1DecodeError> {
+        let mut results = Vec::new();
+        let mut rest = input;
+
+        // while there's data to process
+        while !rest.is_empty() {
+            let max_depth = ad.max_frames as usize;
+
+            if results.len() >= max_depth {
+                return Err(Asn1DecodeError::MaxFrames);
+            }
+
+            let res = parse_ber_recursive(rest, max_depth);
+
+            match res {
+                Ok((new_rest, obj)) => {
+                    results.push(obj);
+
+                    rest = new_rest;
+                }
+                // If there's an error, bail
+                Err(_) => {
+                    break;
+                }
```

Comment:
I understood it as the decoding would happen on ever packet? If not, is it only on `content` matches? (I'm not super familiar with suricata's processing yet)

Reply:
I'm not sure what you mean here. If the function is called for every packet, and you want to ignore decoding errors ("if it's not BER, just ignore ignore packet"), the the code is fine.
I just wanted to be sure you are not silently ignoring errors here, so wondering the initial intent of the code.

Reply:
Right, yes. The intention when writing this code was to ignore decoding errors. My assumption was that this code-path would get hit for every packet and didn't want to log something out, I could be wrong though. Thoughts @victorjulien ?

Reply:
Wonder if we should add a new option to the detection keyword to match on failure to decode. OTOH since this just looks at packet payloads (not reassembled stream) it might fail in benign cases when the data spans multiple packets.

Side thought: maybe its worth making sure we can point this inspection at other buffers than just the packet payload.

Reply:
Nice, if some community tickets were created for these, I know some folks who _may_ be interested in implementing this! ;)

Given that, I'll leave it as a silent fail